extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/configproperties/ConfigPropertiesBuildStep.java

```diff
@@ -0,0 +1,97 @@
+package io.quarkus.arc.deployment.configproperties;
+
+import java.lang.reflect.Modifier;
+import java.util.Collection;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import org.jboss.jandex.AnnotationInstance;
+import org.jboss.jandex.AnnotationValue;
+import org.jboss.jandex.ClassInfo;
+import org.jboss.jandex.IndexView;
+
+import io.quarkus.arc.deployment.GeneratedBeanBuildItem;
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.ApplicationIndexBuildItem;
+import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
+import io.quarkus.deployment.builditem.GeneratedClassBuildItem;
+import io.quarkus.gizmo.ClassCreator;
+import io.quarkus.gizmo.ClassOutput;
+
+public class ConfigPropertiesBuildStep {
+
+    @BuildStep
+    void setup(CombinedIndexBuildItem combinedIndex,
+            ApplicationIndexBuildItem applicationIndex,
+            BuildProducer<GeneratedClassBuildItem> generatedClasses,
+            BuildProducer<GeneratedBeanBuildItem> generatedBeans) {
+        IndexView index = combinedIndex.getIndex();
+        Collection<AnnotationInstance> instances = index.getAnnotations(DotNames.CONFIG_PROPERTIES);
+        if (instances.isEmpty()) {
+            return;
+        }
+
+        ClassOutput beansClassOutput = new ClassOutput() {
+            @Override
+            public void write(String name, byte[] data) {
+                generatedBeans.produce(new GeneratedBeanBuildItem(name, data));
+            }
+        };
+
+        ClassOutput nonBeansClassOutput = new ClassOutput() {
+            @Override
+            public void write(String name, byte[] data) {
+                generatedClasses.produce(new GeneratedClassBuildItem(true, name, data));
+            }
+        };
+
+        /*
+         * We generate CDI producer bean containing one method for each of the @ConfigProperties
+         * instances we encounter
+         */
+
+        ClassCreator producerClassCreator = ClassCreator.builder().classOutput(beansClassOutput)
+                .className("io.quarkus.config.ConfigPropertiesProducer")
```

Comment:
So we generate a bean that declares a producer for each `@ConfigProperties`, right? In theory, we could register custom beans instead but it really depends on the logic you're adding to the producer methods. I'll take a look at the generated class...

Reply:
Yes exactly. The producer instantiates the class by pulling data from MP Config and setting the proper fields or calling setters

Reply:
Also keep in mind that works for both classes and interfaces as config properties holders

Reply:
> Also keep in mind that works for both classes and interfaces as config properties holders

Hm, what do you mean? A generated producer looks like this:

```java
@Produces
public TypicalInterfaceConfigPropertiesTest$DummyProperties produceTypicalInterfaceConfigPropertiesTest$DummyProperties(
			final Config config) {
		return (TypicalInterfaceConfigPropertiesTest$DummyProperties) new TypicalInterfaceConfigPropertiesTest$DummyPropertiesQuarkusImpl(
				config);
}
```

The injected `Config` could be easily obtained using `Arc.container().instance(Config.class).get()`.

On the other hand, having a generated `ConfigPropertiesProducer` is fine. Maybe just use `@Singleton` instead of `@ApplicationScoped` as we don't need client proxies here.

Reply:
I see what you mean yes.

TBH, that is what I tried initially for interfaces, but then I changed it because I had to make `Config` and unremovable bean. So the current solution seemed cleaner and more consisent with class handling case.

WDYT?

Reply:
Yes, makes sense. I'd keep the current solution but change the scope of `ConfigPropertiesProducer`.

Reply:
:+1: . What is the proper scope for this case?

Reply:
:+1: . What is the proper scope for this case?

Reply:
Oops, never mind, I see you mentioned `@Singleton` earlier :)